[OKTA-459397] Okta Verify (Push/OTP) integration guide (.NET SDK) (#3118)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/authenticators-okta-verify/main/aspnet/enrollmentotherintegrationsteps.md

> #### 1. Initiate sign-in and return a list of authenticators

Reviewer: ```suggestion
#### 1. Initiate sign-in flow and return a list of authenticators
```

Author: fixed

The user signs in with a username and password, and then chooses Okta Verify from a list of authenticators. This is covered in the earlier [Shared Code](#initiate-sign-in-and-return-a-list-of-authenticators) section.

#### 2 - 3. Retrieve QR code, display QR code

> When the user selects Okta Verify and clicks **Submit**, the form posts back to the `SelectAuthenticatorAsync` method. In Enrollment Flow, a QR code is returned to be displayed to the user, along with a link to select an alternative enrollment method.

Reviewer: ```suggestion
When the user selects Okta Verify and clicks **Submit**, the form posts back to the `SelectAuthenticatorAsync` method. In Enrollment Flow, a QR code is returned and appears for the user, along with a link to select an alternative enrollment method.
```

The code for this is shared with Steps 2 and 3 of [Integrate Enrollment Using QR code](#integrate-enrollment-using-qr-code).

#### 4. Display a list of other enrollment channels

> Build a page to display the other enrollment channels for Okta Verify. For example, in the sample application, a new `OktaVerifySelectEnrollmentChannelModel` is populated from the `oktaVerifyAuthenticator` object stored in session in the previous step.

Reviewer: ```suggestion
Build a page to display the other enrollment channels for Okta Verify. For example, in the sample application, a new `OktaVerifySelectEnrollmentChannelModel` is populated from the `oktaVerifyAuthenticator` object that is stored in the session in the previous step.
```

Author: fixed

```csharp
[HttpGet]
public ActionResult SelectEnrollmentChannel()
{
   var oktaVerifyAuthenticator = (IAuthenticator)Session["oktaVerifyAuthenticator"];
   var selectEnrollmentChannelModel = new OktaVerifySelectEnrollmentChannelModel(
      oktaVerifyAuthenticator);
   return View(selectEnrollmentChannelModel);
}
```

> The `OktaVerifySelectEnrollmentChannelModel` constructor retains the authenticator ID and a list of available enrollment channels that are not QR code.

Reviewer: ```suggestion
The `OktaVerifySelectEnrollmentChannelModel` constructor retains the authenticator ID and a list of available enrollment channels that aren't the QR code.
```

Author: fixed

      FormMethod.Post, new { role = "form" }))
    {
        @Html.AntiForgeryToken()
        @Html.HiddenFor(m => m.AuthenticatorId)
        <h4>More options</h4>
        <div>Which option do you want to try?</div>

        <div>
            <div>
                @foreach (var channel in Model.ChannelTypes)
                {
                <div>
                    <label>
                        @Html.RadioButtonFor(m => m.SelectedChannel, channel)
                        @channel
                    </label>
                </div>
                }
            </div>
        </div>
        <div>
            <div>
                <input type="submit" value="Next" />
            </div>
        </div>
        @Html.ValidationSummary("", new { @class = "text-danger" })
    }
</div>
```

The available channels are listed, as shown in the following screenshot.

<div class="common-image-format">

![A list showing Okta Verify enrollment channels available for use](/img/authenticators/dotnet-authenticators-okta-verify-enrollment-list-of-other-enrollment-options.png "A list of the available Okta Verify enrollment channels")

</div>

> #### 5. Initiate Email Enrollment On Server-Side

Reviewer: ```suggestion
#### 5. Initiate email enrollment on server-side
```

```csharp
[HttpPost]
public async Task<ActionResult> SelectEnrollmentChannel(OktaVerifySelectEnrollmentChannelModel model)
{
    var oktaVerifyAuthenticator = (IAuthenticator)Session["oktaVerifyAuthenticator"];
    var idxContext = (IIdxContext)Session["idxContext"];
    var enrollOktaVerifyOptions = new EnrollOktaVerifyAuthenticatorOptions
    {
        Channel = model.SelectedChannel,
        AuthenticatorId = model.AuthenticatorId
    };

    await _idxClient.SelectEnrollAuthenticatorAsync(enrollOktaVerifyOptions, idxContext);
```

Send the user to a page specific to the email channel that asks them for their email address.

```csharp
    switch (model.SelectedChannel)
    {
        case "email":
            return View("EnrollWithEmail");
        case "sms":
            return View("EnrollWithPhoneNumber", new OktaVerifyEnrollWithPhoneNumberModel { CountryCode = "+1" });
    }

    ModelState.AddModelError("SelectedChannel", new ArgumentException($"Unrecognized Okta Verify channel: {model.SelectedChannel}"));
    return View(model);
}
```

> #### 6. Display Prompt for User's Email Address

Reviewer: ```suggestion
#### 6. Display prompt for user's email address
```

Author: fixed

> Build a page to display a prompt for the user's email address. For example, in the sample app, we have:

Reviewer: ```suggestion
Build a page that displays a prompt for the user's email address. For example, in the sample app, we have:
```

```csharp
[HttpPost]
public async Task<ActionResult> EnrollWithEmail(OktaVerifyEnrollWithEmailModel emailModel)
{
    var idxContext = (IIdxContext)Session["idxContext"];
    var okaVerifyEnrollWithEmailOptions = new EnrollOktaVerifyAuthenticatorOptions
    {
        Channel = "email",
        Email = emailModel.Email,
    };

    _ = await _idxClient.EnrollAuthenticatorAsync(okaVerifyEnrollWithEmailOptions, idxContext);
    var oktaVerifyEnrollModel = new OktaVerifyEnrollPollModel()
    {
        Message = "An activation link was sent to your email, use it to complete Okta Verify enrollment."
    };

    return View("EnrollPoll", oktaVerifyEnrollModel);
}
```

#### 8. Send an email with Okta Verify activation link

After the user submits their email, Okta sends an activation link to that address.

<div class="common-image-format">

![An email that contains the Okta Verify Push activation link](/img/authenticators/dotnet-authenticators-okta-verify-enrollment-email-text.png "An email containing the activation link")

</div>

#### 9. Prompt user to check email and start polling

> Build a page to display the prompt for the user to check their email while starting to poll Okta for completion.  In the sample application, a new `OktaVerifyEnrollPollModel` is populated in the previous step containing a message to the user to look in their email for the next step.

Reviewer: ```suggestion
Build a page that displays the prompt for the user to check their email while starting to poll Okta for completion.  In the sample application, a new `OktaVerifyEnrollPollModel` is populated in the previous step that contains a message to the user to look in their email for the next step.
```

Author: fixed

> This ViewModel is then consumed in a Razor page. This page contains two important sections. First, the form that contains the QR code and drop-down of other enrollment options.

Reviewer: ```suggestion
This ViewModel is then consumed in a Razor page. This page contains two important sections. First, the form that contains the QR code and dropdown list of other enrollment options.
```

```razor
<div>
    <h4>Set up Okta Verify</h4>
    <ol>
        <li>On your mobile device, download the Okta Verify app from the App Store (iPhone and iPad) or Google Play (Android devices).</li>
        <li>@Model.Message</li>
    </ol>
    <hr />
</div>
```

> Second, the JavaScript `poll()` function that will continue to poll an endpoint in the app on the server-side (in this case, `/OktaVerify/EnrollPoll`) until an indication is given that Okta Verify has been enrolled successfully on the server-side.

Reviewer: ```suggestion
Second, the JavaScript `poll()` function that continues to poll an endpoint in the app on the server-side (in this case, `/OktaVerify/EnrollPoll`) until an indication is given that Okta Verify is enrolled successfully on the server-side.
```

Author: fixed

```js
<script>
    function poll() {
        fetch('@Model.PollEndpoint')
            .then(response => response.json())
            .then(data => {
                if (data.ContinuePolling) {
                    setTimeout(poll, data.Refresh);
                } else {
                    window.location.href = data.Next;
                }
            });
    }

    setTimeout(poll, @Model.RefreshInterval);
</script>
```

> The polling logic and code is common across all the flows that use it and is covered [here](#polling-okta).

Reviewer: ```suggestion
The polling logic and code is common across all the flows that use it and is covered in the [Polling Okta](#polling-okta) section.
```

> #### 10. Click on the link in the email and complete Okta Verify setup

Reviewer: ```suggestion
#### 10. Click the link in the email and complete Okta Verify setup
```

Author: fixed

> The user opens the email on their mobile device and clicks Activate Okta Verify Push, which sends them to Okta Verify to complete the account setup.

Reviewer: ```suggestion
The user opens the email on their mobile device and taps **Activate Okta Verify Push** that sends them to Okta Verify to complete the account setup.
```

<div class="common-image-format">

![A prompt for a user to install Okta Verify and activate it](/img/authenticators/dotnet-authenticators-okta-verify-enrollment-email-prompt-user-to-check-email.png "An activation prompt for Okta Verify")

</div>

> #### 11. Exit Polling

Reviewer: ```suggestion
#### 11. Exit polling
```

Author: fixed

> After the user completes the setup and finishes enrolling the account, the next time `EnrollPoll()` is called, it returns a status of `Success` along with access and ID tokens. The page exits the polling and redirects the user to the default home page for the signed-in user.

Reviewer: ```suggestion
After the user completes the setup and finishes enrolling their account, the next time `EnrollPoll()` is called, it returns a status of `Success` along with access and ID tokens. The page exits the polling and redirects the user to the default home page for the signed-in user.
```
